precompile/allowlist/allowlist.go

```diff
@@ -160,20 +156,67 @@ func CreateAllowListPrecompile(precompileAddr common.Address) contract.StatefulP
 	// Construct the contract with no fallback function.
 	allowListFuncs := CreateAllowListFunctions(precompileAddr)
 	contract, err := contract.NewStatefulPrecompileContract(nil, allowListFuncs)
-	// TODO Change this to be returned as an error after refactoring this precompile
-	// to use the new precompile template.
 	if err != nil {
 		panic(err)
 	}
 	return contract
 }
 
 func CreateAllowListFunctions(precompileAddr common.Address) []*contract.StatefulPrecompileFunction {
-	setAdmin := contract.NewStatefulPrecompileFunction(setAdminSignature, createAllowListRoleSetter(precompileAddr, AdminRole))
-	setManager := contract.NewStatefulPrecompileFunctionWithActivator(setManagerSignature, createAllowListRoleSetter(precompileAddr, ManagerRole), contract.IsDUpgradeActivated)
-	setEnabled := contract.NewStatefulPrecompileFunction(setEnabledSignature, createAllowListRoleSetter(precompileAddr, EnabledRole))
-	setNone := contract.NewStatefulPrecompileFunction(setNoneSignature, createAllowListRoleSetter(precompileAddr, NoRole))
-	read := contract.NewStatefulPrecompileFunction(readAllowListSignature, createReadAllowList(precompileAddr))
+	var functions []*contract.StatefulPrecompileFunction
+
+	type precompileFn struct {
+		fn        contract.RunStatefulPrecompileFunc
+		activator contract.ActivationFunc
+	}
+
+	abiFunctionMap := map[string]precompileFn{
+		getAllowListFunctionSelector(AdminRole): {
+			fn: createAllowListRoleSetter(precompileAddr, AdminRole),
+		},
+		getAllowListFunctionSelector(EnabledRole): {
+			fn: createAllowListRoleSetter(precompileAddr, EnabledRole),
+		},
+		getAllowListFunctionSelector(NoRole): {
+			fn: createAllowListRoleSetter(precompileAddr, NoRole),
+		},
+		"readAllowList": {
+			fn: createReadAllowList(precompileAddr),
+		},
+		getAllowListFunctionSelector(ManagerRole): {
+			fn:        createAllowListRoleSetter(precompileAddr, ManagerRole),
+			activator: contract.IsDUpgradeActivated,
+		},
+	}
+
+	for name, function := range abiFunctionMap {
+		method, ok := AllowListABI.Methods[name]
+		if !ok {
+			panic(fmt.Errorf("given method (%s) does not exist in the ABI", name))
+		}
+		var spFn *contract.StatefulPrecompileFunction
+		if function.activator != nil {
+			spFn = contract.NewStatefulPrecompileFunctionWithActivator(method.ID, function.fn, function.activator)
+		} else {
+			spFn = contract.NewStatefulPrecompileFunction(method.ID, function.fn)
+		}
+		functions = append(functions, spFn)
+	}
+
+	return functions
+}
 
-	return []*contract.StatefulPrecompileFunction{setAdmin, setManager, setEnabled, setNone, read}
+func getAllowListFunctionSelector(role Role) string {
```

Comment:
No, only the hash of the role is being stored in on the chain. 

I thought this would be too specific for role struct as we only need this in the contract code. But I can move it to Role struct with a receiver method.

Reply:
I moved it to the role struct. The role is an alias to `common.Hash` and I'm not quite sure we can get rid of all switch cases (CanModify, IsEnabled etc)